### ISSUE TITLE
`nvim_buf_set_lines` got error when updating hidden buffer

### ISSUE THREAD

**Issue by chemzqm** [neovim/neovim#6072]
- `nvim --version`: NVIM v0.2.0-532-gf61e4da1
- Vim (version: ) behaves differently? not exist for vim
- Operating system/version: MacOS
- Terminal name/version: iTerm2
- `$TERM`: xterm-256

### Actual behaviour

Got error E315: `ml_get: invalid lnum: 21`

### Expected behaviour

No error thrown.

### Steps to reproduce using `nvim -u NORC`

Source vim code like:
```
function! SetLines(...)
  let lines = []
  call add(lines, strftime('%c'))
  call nvim_buf_set_lines(g:nr, 0, len(lines), v:false, lines)
endfunction

edit /tmp/test
let g:nr = bufnr('%')

call timer_start(1000, 'SetLines', {'repeat': -1})
```

It should works at begining, but an error would thrown after switch to another buffer and move the cursor.

**Comment by hardenedapple** [comment]
This seems to be the problem I fixed in PR #6019 -- it works on my machine with that patch applied.

**Comment by chemzqm** [comment]
@hardenedapple thanks, I can confirm that PR works.
